### DISCUSSION TITLE
Locking in middleware

### DISCUSSION THREAD

**Discussion by gitzko** [serilog-contrib/serilog-enrichers-globallogcontext#51] [Q&A] [answered]
Is GlobalLogContext.Lock() required in a constructor in an ASP.NET Core middleware?
Also, is there a better place(than middleware) to set one log property for the whole project per request?

**Comment by augustoproiete** [comment, marked as answer]
Hey @gitzko,`Lock` (or `LockAsync`) is not needed in a constructor of an ASP .NET Core middleware because the middleware is constructed only once per application lifetime, so there's no multiple threads writing to the `GlobalLogContext`.

As for your second question, It's hard to answer without knowing more details of your app, but fundamentally I'd recommend that you set the property at the closest place where you have the value for it... If the middleware constructor is where you first have the value you want to set globally, then that seems like the right place for it.

**Reply by gitzko** [reply]
Hi Augusto, thanks for answering. 
I ended up placing the GlobalLogContext push in the middleware's InvokeAsync method, and I guess there I have to add the lock.

**Reply by augustoproiete** [reply]
@gitzko Indeed, the `InvokeAsync` method is called for all requests, so you would need the lock there - that said, if the value you need is only available per every request, then chances are that you shouldn't be using `GlobalLogContext` and instead should be using `LogContext` (which is isolated for each request, and no lock required)

See https://github.com/serilog/serilog/wiki/Enrichment

**Reply by gitzko** [reply]
But this doesn't seem to work with Serilog's UseSerilogRequestLogging, which is a middleware by itself, and we're using it.

**Reply by gitzko** [reply]
Except it does, when I reorder the app configuration calls. Thanks!
